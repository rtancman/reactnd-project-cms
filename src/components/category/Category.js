import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { ListContent } from '../layout/List'
import NotFoundPage from '../pages/NotFoundPage'

class Category extends Component {
  static propTypes = {
    categoryId: PropTypes.string.isRequired,
    categories: PropTypes.object.isRequired,
    posts: PropTypes.object.isRequired,
  }

  render() {
    let content = ''
    const { categoryId, categories, posts } = this.props
    const validCategory = categories.items.filter((c) => c.path === categoryId)

    if (validCategory.length !== 1) {
      return(<NotFoundPage/>)
    }

    const categoryPosts = posts.items.filter((p) => p.category === categoryId)

    if ( categoryPosts.length > 0 ) {
      content = (<ListContent title='Posts' items={categoryPosts} />)
    }

    return (
      <div className='Category'>
        <div className='content'>
          <div className='content__head'>
            <h1 className='content__title'>{ categoryId }</h1>
            <hr />
          </div>
          { content }
        </div>
      </div>
    )
  }
}

const mapStateToProps = ({ posts, categories }) => {
  return {
    posts,
    categories,
  }
}

export default connect(mapStateToProps)(Category)
