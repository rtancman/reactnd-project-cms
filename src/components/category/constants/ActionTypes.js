export const REQUEST_CATEGORIES = 'REQUEST_CATEGORIES'
export const RECEIVE_CATEGORIES = 'RECEIVE_CATEGORIES'
export const INVALIDATE_CATEGORIES = 'INVALIDATE_CATEGORIES'

export const REQUEST_CATEGORY_POSTS = 'REQUEST_CATEGORY_POSTS'
export const RECEIVE_CATEGORY_POSTS = 'RECEIVE_CATEGORY_POSTS'
export const INVALIDATE_CATEGORY_POSTS = 'INVALIDATE_CATEGORY_POSTS'
