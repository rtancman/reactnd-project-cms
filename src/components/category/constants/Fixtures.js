export const categoriesMock = {
  categories: [{
    "name": "react",
    "path": "react"
    },
    {
      "name": "redux",
      "path": "redux"
    },
    {
      "name": "udacity",
      "path": "udacity"
    }
  ]
}

export const categoryPostsMock = [
  {
      "id": "8xf0y6ziyjabvozdd253nd",
      "timestamp": 1467166872634,
      "title": "Udacity is the best place to learn React",
      "body": "Everyone says so after all.",
      "author": "thingtwo",
      "category": "react",
      "voteScore": 6,
      "deleted": false,
      "commentCount": 2
  }
]