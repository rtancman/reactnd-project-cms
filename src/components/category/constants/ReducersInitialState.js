export const initialCategoriesState = {
  isFetching: false,
  didInvalidate: false,
  items: [],
}