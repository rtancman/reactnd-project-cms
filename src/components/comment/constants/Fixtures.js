export const commentMock = {
  "id": "AAAAb033-81ca-4bea-8f51-237ac1694d70",
  "timestamp": 1536967930576,
  "body": "Redux is amazing!\nRedux is amazing!\nRedux is amazing!\nRedux is amazing!\nRedux is amazing!\nRedux is amazing!\nRedux is amazing!Redux is amazing!\n\n:)",
  "author": "Raffael Tancman",
  "parentId": "e91ab033-81ca-4bea-8f51-237ac1694d70",
  "voteScore": 1,
  "deleted": false,
  "parentDeleted": false
}