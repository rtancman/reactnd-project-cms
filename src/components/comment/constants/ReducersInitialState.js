export const initialCreateCommentState = {
  isFetching: false,
  didInvalidate: false,
  created: false
}