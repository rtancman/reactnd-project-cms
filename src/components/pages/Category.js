import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Grid from '@material-ui/core/Grid';


const styles = theme => ({
  root: {
    flexGrow: 1,
  }
});

const Category = (props) => {
  const { classes } = props;

  return (
    <div className={classes.root}>
      <h1>Category</h1>
      <Grid container spacing={24}>
      </Grid>
    </div>
  )
}

Category.propTypes = {
  classes: PropTypes.object.isRequired,
}

export default withStyles(styles)(Category);