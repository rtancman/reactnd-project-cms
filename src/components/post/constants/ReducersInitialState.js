export const initialListPostState = {
  isFetching: false,
  didInvalidate: false,
  items: [],
}

const post = {
  id: '',
  timestamp: 0,
  title: '',
  body: '',
  author: '',
  category: '',
  voteScore: 0,
  deleted: false,
  commentCount: 0
}

export const initialPostState = {
  isFetching: false,
  didInvalidate: false,
  content: {},
  comment: {}
}

export const initialCreatePostState = {
  isFetching: false,
  didInvalidate: false,
  created: false
}

export const initialRemovePostState = {
  isFetching: false,
  didInvalidate: false,
  removed: false,
  id: '',
}