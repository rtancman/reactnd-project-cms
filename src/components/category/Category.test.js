import React from 'react'
import { shallow, mount } from 'enzyme';
import configureStore from 'redux-mock-store'
import fetchMock from 'fetch-mock'
import { MemoryRouter } from 'react-router';
import thunk from 'redux-thunk'
import Category from './Category'
import { Provider } from 'react-redux';
import { listCategoriesUrl, listPostsUrl } from '../../api/cms';
import { listPostsMock } from '../post/constants/Fixtures'
import { initialListPostState } from '../post/constants/ReducersInitialState'
import { initialCategoriesState } from './constants/ReducersInitialState'
import { categoriesMock } from './constants/Fixtures'


describe('Category Component', () => {
  fetchMock.getOnce(listCategoriesUrl, { body: {categories: categoriesMock.categories} })
  fetchMock.getOnce(listPostsUrl, { body: listPostsMock })
  const middlewares = [thunk]
  const mockStore = configureStore(middlewares)
  const categoryId = categoriesMock.categories[0].path
  const fetchData = jest.fn(() => Promise.resolve({body:{}}))
  let store
  
  describe('when component renders', () => {
    beforeEach(() => {
      store = mockStore({
        categories: {...initialCategoriesState, items: categoriesMock.categories},
        posts: {...initialListPostState, items: listPostsMock},
      })
    })

    it('without crashing', () => {
      const props = {
        store: store,
        categories: {},
        posts: {},
        categoryId: categoryId,
      }
      let wrapper = shallow(<Category {...props} />)

      expect(wrapper).toMatchSnapshot();
    })

    afterEach(() => {
      fetchMock.reset()
      fetchMock.restore()
    })

    it('displays the NotFoundPage component when request category posts', () => {
      const props = {
        categoryId: '1111',
      }
      const root = mount(
        <Provider store={store}>
          <MemoryRouter>
            <Category {...props} />
          </MemoryRouter>
        </Provider>
      );

      const wrapper = root.find('Category')
      expect(wrapper.length).toBe(1)
      expect(wrapper.find('NotFoundPage').length).toBe(1);
      expect(wrapper.find('ListContent').length).toBe(0);
    })

    it('displays the ListContent component when request category posts', () => {
      const props = {
        categoryId: categoryId,
      }
      const root = mount(
        <Provider store={store}>
          <MemoryRouter>
            <Category {...props} />
          </MemoryRouter>
        </Provider>
      );
      
      let wrapper = root.find('Category')
      expect(wrapper.length).toBe(1)
      wrapper.instance().setState({isFetching: false})
      root.update()
      wrapper = root.find('Category')
      expect(wrapper.find('NotFoundPage').length).toBe(0);
      expect(wrapper.find('ListContent').length).toBe(1);
    })

    it('calls componentDidMount() lifecycle method', () => {
      const componentDidMountSpy = jest.spyOn(Category.prototype, 'componentDidMount');
      const props = {
        store: store,
        fetchData: fetchData,
        items: [],
        categoryId: categoryId,
      }
      let wrapper = shallow(<Category {...props} />)

      expect(Category.prototype.componentDidMount).toHaveBeenCalledTimes(1)
      componentDidMountSpy.mockRestore()
    })
  })
})