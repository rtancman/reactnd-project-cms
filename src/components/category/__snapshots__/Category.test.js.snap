// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category Component when component renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(Category)
    categories={Object {}}
    categoryId="react"
    posts={Object {}}
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "categories": Object {
        "items": Array [
          Object {
            "name": "react",
            "path": "react",
          },
          Object {
            "name": "redux",
            "path": "redux",
          },
          Object {
            "name": "udacity",
            "path": "udacity",
          },
        ],
      },
      "categoryId": "react",
      "dispatch": [Function],
      "posts": Object {
        "didInvalidate": false,
        "isFetching": false,
        "items": Array [
          Object {
            "author": "thingtwo",
            "body": "Everyone says so after all.",
            "category": "react",
            "commentCount": 2,
            "deleted": false,
            "id": "8xf0y6ziyjabvozdd253nd",
            "timestamp": 1467166872634,
            "title": "Udacity is the best place to learn React",
            "voteScore": 6,
          },
          Object {
            "author": "thingone",
            "body": "Just kidding. It takes more than 10 minutes to learn technology.",
            "category": "redux",
            "commentCount": 0,
            "deleted": false,
            "id": "6ni6ok3ym7mf1p33lnez",
            "timestamp": 1468479767190,
            "title": "Learn Redux in 10 minutes!",
            "voteScore": -5,
          },
        ],
      },
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "categories": Object {
          "items": Array [
            Object {
              "name": "react",
              "path": "react",
            },
            Object {
              "name": "redux",
              "path": "redux",
            },
            Object {
              "name": "udacity",
              "path": "udacity",
            },
          ],
        },
        "categoryId": "react",
        "dispatch": [Function],
        "posts": Object {
          "didInvalidate": false,
          "isFetching": false,
          "items": Array [
            Object {
              "author": "thingtwo",
              "body": "Everyone says so after all.",
              "category": "react",
              "commentCount": 2,
              "deleted": false,
              "id": "8xf0y6ziyjabvozdd253nd",
              "timestamp": 1467166872634,
              "title": "Udacity is the best place to learn React",
              "voteScore": 6,
            },
            Object {
              "author": "thingone",
              "body": "Just kidding. It takes more than 10 minutes to learn technology.",
              "category": "redux",
              "commentCount": 0,
              "deleted": false,
              "id": "6ni6ok3ym7mf1p33lnez",
              "timestamp": 1468479767190,
              "title": "Learn Redux in 10 minutes!",
              "voteScore": -5,
            },
          ],
        },
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
